0xProject/0x-monorepo
# Add support for TypeScript project references

## Description



[Project references](https://www.typescriptlang.org/docs/handbook/project-references.html) are a new feature in TypeScript 3.0 which can enable faster, more efficient builds for projects which are composed of many packages (e.g. a monorepo like our own).

Getting this to work is a lot harder than it may seem on the surface. I needed to update a lot of __tsconfig.json__ and __package.json__ files or else certain packages would break in subtle and surprising ways (for example, in some packages I had to add `"rootDir": "./src"`).

There are two new commands in __package.json__ at the monorepo root.

1. `yarn build:ts` will build build all the necessary TypeScript files for all packages in the correct order based on their dependencies. It is much smarter about which files need to be recompiled and is pretty much instantaneous if no TypeScript source files have changed. However, it does not run any `pre_build` steps. This means that typically you will need to run `yarn build` the first time you clone the monorepo or e.g., if you made any changes to contracts.
2. `yarn watch:ts` is like `yarn build:ts` but actively and continuously watches all TypeScript files across all packages, recompiling every package and those that depend on it as needed.

Everything else works pretty much the same as before but will be a little bit more efficient in some cases.

As an added bonus: with this change "Go to Definition" in VS Code will take you to the TypeScript source code with the definition instead of the old behavior of taking you to the __.d.ts__ file 🎉 


## Testing instructions



Run `yarn build` and `yarn test`. Test out `yarn build:ts` and `yarn watch:ts` to make sure they work as expected.

## Types of changes





* New feature (non-breaking change which adds functionality)



## Checklist:
